### PR TITLE
Docstring for ModelChainResult

### LINKED ISSUE #1126: Improve documentation of ModelChainResult class
The RTD page gives a nice list of fields, but is otherwise quite bare. We should add docstrings describing each field. 

This is follow on work from #1076.

### PR DESCRIPTION
**Pull request by cwhanse**
- [x] Closes #1126
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] ~~Tests added~~
 - [x] ~~Updates entries to [`docs/sphinx/source/api.rst`](https://github.com/pvlib/pvlib-python/blob/master/docs/sphinx/source/api.rst) for API changes.~~
 - [x] ~~Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/master/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).~~
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

### CODE REVIEW

File: pvlib/modelchain.py [hunk 1]
```diff
@@ -260,27 +260,99 @@ class ModelChainResult:
...
+    """Solar position in a DataFrame containing columns ``'apparent_zenith'``,
+    ``'zenith'``, ``'apparent_elevation'``, ``'elevation'``, ``'azimuth'``
+    (all in degrees), with possibly other columns depending on the solar
+    position method; see :py:meth:`~pvlib.solarposition.get_solarposition`
+    for details."""
+
     airmass: Optional[pd.DataFrame] = field(default=None)
+    """Air mass in a DataFrame containing columns ``'airmass_relative'``,
+    ``'airmass_absolute'`` (unitless); see
+    :py:meth:`~pvlib.location.get_airmass` for details."""
```
Comment (kanderso-nrel) [inline pvlib/modelchain.py:272]:
```suggestion
    :py:meth:`~pvlib.location.Location.get_airmass` for details."""
```

**Review by kanderso-nrel — COMMENTED** [review]
`PerArray` makes an undesirable appearance in the list of attributes ([link](https://pvlib-python--1233.org.readthedocs.build/en/1233/generated/pvlib.modelchain.ModelChainResult.html)); should we rename it to `_PerArray`?

File: pvlib/modelchain.py [hunk 2]
```diff
@@ -260,27 +260,99 @@ class ModelChainResult:
...
+    see :py:func:`~pvlib.pvsystem.singlediode` for details.
+    """
+
     dc_ohmic_losses: Optional[PerArray[pd.Series]] = field(default=None)
+    """Series (or tuple of Series, one for each array) containing DC ohmic
+    loss (W) calculated by ``ModelChain.dc_ohmic_model``.
+    """
+
     losses: Optional[Union[pd.Series, float]] = field(default=None)
+    """Series (or tuple of Series, one for each array) containing DC loss
```
Comment (wholmgren) [inline pvlib/modelchain.py:343]: never a tuple

File: pvlib/modelchain.py [hunk 3]
```diff
@@ -260,27 +260,99 @@ class ModelChainResult:
...
+    as a fraction of DC power, as calculated by ``ModelChain.losses_model``.
+    """
 
     weather: Optional[PerArray[pd.DataFrame]] = None
+    """DataFrame (or tuple of DataFrame, one for each array) contains a
+    copy of the input weather data.
+    """
+
     times: Optional[pd.DatetimeIndex] = None
+    """DatetimeIndex (or tuple of DatetimeIndex, one for each array) contains
```
Comment (wholmgren) [inline pvlib/modelchain.py:353]: never a tuple

**Review by wholmgren — APPROVED** [review]
this is great!

**Comment by wholmgren** [comment]
> PerArray makes an undesirable appearance in the list of attributes (link); should we rename it to _PerArray?

type linters complain about the line `_T = TypeVar('T')` - they want it to be `T = TypeVar('T')`. I assume that would mean the generic type var would also show up in the docs. So maybe there is a way to exclude the attributes? Or move them somewhere else?

**Comment by kanderso-nrel** [comment]
> 
> 
> > PerArray makes an undesirable appearance in the list of attributes (link); should we rename it to _PerArray?
> 
> type linters complain about the line `_T = TypeVar('T')` - they want it to be `T = TypeVar('T')`. I assume that would mean the generic type var would also show up in the docs. So maybe there is a way to exclude the attributes? Or move them to the module scope?

Moving to module scope seems to solve the problem, although I did not get mypy's opinion about it.  I looked around a bit for how other packages document dataclasses and came up mostly short -- this is the only example I found, which uses the standard approach of docstring sections: https://github.com/scipy/scipy/blob/701ffcc8a6f04509d115aac5e5681c538b5265a2/scipy/stats/_relative_risk.py#L21

**Comment by wholmgren** [comment]
I'd also like to see the type hints removed from the class signature. Nobody is going to directly instantiate this class, but all that text could be distracting/intimidating. My reading is that we just need to add `autodoc_typehints = 'none'` to `conf.py` (see [sphinx docs](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_typehints)) but that isn't working for me.

**Comment by cwhanse** [comment]
> I'd also like to see the type hints removed from the class signature. Nobody is going to directly instantiate this class, but all that text could be distracting/intimidating. My reading is that we just need to add `autodoc_typehints = 'none'` to `conf.py` (see [sphinx docs](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_typehints)) but that isn't working for me. Might be specific to the dataclass init - maybe find a way to skip the signature entirely?

I agree that text is distracting, but I have no idea how to suppress it either.

**Comment by cwhanse** [comment]
@wholmgren hints to controlling a dataclass docstring [here](https://stackoverflow.com/questions/51125415/how-do-i-document-a-constructor-for-a-class-using-python-dataclasses)?

File: pvlib/modelchain.py [hunk 4]
```diff
@@ -245,12 +245,15 @@ def get_orientation(strategy, **kwargs):
     return surface_tilt, surface_azimuth
 
 
+# Type for fields that vary between arrays
+_T = TypeVar('T')
+
+
+_PerArray = Union[_T, Tuple[_T, ...]]
```
Comment (wholmgren) [inline pvlib/modelchain.py:252]: No underscore on either of these. Type variables should be public so other libraries that build on pvlib can inspect them and benefit from them.

**Review by wholmgren — COMMENTED** [review]
Let's not worry about the signature in this pr.

**Review by wholmgren — APPROVED** [review]
merge when green.

**Pull request merged**